File: Linked List /C++/1. 141. Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// Using naive approach
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *temp = head;
        unordered_set <ListNode*> s;
        while(temp != NULL)
        {
            if(s.find(temp)!= s.end())
            {
                return true;
            }
            else
            {
                s.insert(temp);
                temp = temp ->next;
            }
        }
        return false;
    }
};

File: Linked List /Python/1.  141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        s = set()   # Create a set which stores visited nodes
        temp = head
        while temp:
            if temp in s:  # if that particular node is present in set "s"
                return True
            else:
                s.add(temp)  # Add a node if it's unseen
            temp = temp.next
        return False
